tests/json_infra/conftest.py

```diff
@@ -243,15 +234,70 @@ def pytest_sessionstart(session: Session) -> None:  # noqa: U100
                 )
 
 
+def pytest_sessionstart(session: Session) -> None:  # noqa: U100
+    lock_path = session.config.rootpath.joinpath("tests/fixtures/.lock")
+
+    # use portalocker for mutmut runs
+    if os.environ.get("MUTANT_UNDER_TEST"):
+        shared_lock = portalocker.Lock(lock_path, flags=portalocker.LOCK_SH)
+        shared_lock.acquire()
+        session.stash['mutmut_shared_lock'] = shared_lock
+
+        all_fixtures_ready = all(
+            os.path.exists(props["fixture_path"])
+            for props in TEST_FIXTURES.values()
+        )
+        if not all_fixtures_ready:
+            shared_lock.release()
+            with portalocker.Lock(lock_path, flags=portalocker.LOCK_EX):
+                all_fixtures_ready = all(
+                    os.path.exists(props["fixture_path"])
+                    for props in TEST_FIXTURES.values()
+                )
+                if not all_fixtures_ready:
+                    download_fixtures(session.config.rootpath)
+            shared_lock.acquire()
```

Comment:
Is it possible that between when the exclusive lock is released and when the shared lock is taken that another process acquires the exclusive lock and updates the fixtures to another version?

I think your `while True` approach where you checked every iteration for the conditions was a good approach.

Reply:
got it! added that back. Also added a try-except.